File: frontend/lib/utils/orderItemSort.ts
import type { OrderItem } from "@/types/order";

/**
 * Whether an order item has been packed in full. Items whose `packed_quantity`
 * is below `piece_count * quantity` (including zero) are not fully packed.
 */
export function isOrderItemFullyPacked(item: OrderItem): boolean {
  const totalPieces = (item.piece_count || 1) * item.quantity;
  return (item.packed_quantity ?? 0) >= totalPieces;
}

/**
 * Returns a new array with not-fully-packed items first and fully packed items
 * last. Stable: items inside each group keep their original relative order.
 * Never mutates `items`.
 */
export function sortOrderItemsUnpackedFirst(items: OrderItem[]): OrderItem[] {
  return items
    .map((item, index) => ({ item, index }))
    .sort((a, b) => {
      const groupDiff =
        (isOrderItemFullyPacked(a.item) ? 1 : 0) -
        (isOrderItemFullyPacked(b.item) ? 1 : 0);
      if (groupDiff !== 0) return groupDiff;
      return a.index - b.index;
    })
    .map(({ item }) => item);
}

File: mobile/lib/core/utils/order_item_sort.dart
/// Mirrors `frontend/lib/utils/orderItemSort.ts`.
///
/// Order-summary lists show not-fully-packed items first and fully packed
/// items last, while keeping the original relative order within each group.
library;

import '../../models/models.dart';

/// Whether an order item has been packed in full. Items whose `packedQuantity`
/// is below `pieceCount * quantity` (including zero) are not fully packed.
bool isOrderItemFullyPacked(OrderItem item) {
  final totalPieces = (item.pieceCount ?? 1) * item.quantity;
  return (item.packedQuantity ?? 0) >= totalPieces;
}

/// Returns a new list with not-fully-packed items first and fully packed items
/// last. Stable: items inside each group keep their original relative order.
/// Never mutates [items].
List<OrderItem> sortOrderItemsUnpackedFirst(List<OrderItem> items) {
  final indexed = <({OrderItem item, int index})>[
    for (var i = 0; i < items.length; i++) (item: items[i], index: i),
  ];
  indexed.sort((a, b) {
    final groupA = isOrderItemFullyPacked(a.item) ? 1 : 0;
    final groupB = isOrderItemFullyPacked(b.item) ? 1 : 0;
    if (groupA != groupB) return groupA.compareTo(groupB);
    return a.index.compareTo(b.index);
  });
  return [for (final entry in indexed) entry.item];
}
